// @rakered/errors/lib/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NotFoundError = exports.UserInputError = exports.ForbiddenError = exports.AuthenticationError = void 0;
class AuthenticationError extends Error {
    constructor() {
        super(...arguments);
        this.name = 'AuthenticationError';
        this.code = 401;
    }
}
exports.AuthenticationError = AuthenticationError;
class ForbiddenError extends Error {
    constructor() {
        super(...arguments);
        this.name = 'ForbiddenError';
        this.code = 403;
    }
}
exports.ForbiddenError = ForbiddenError;
class UserInputError extends Error {
    constructor(message, data) {
        super(message);
        this.name = 'UserInputError';
        this.code = 422;
        this.data = data;
    }
}
exports.UserInputError = UserInputError;
class NotFoundError extends Error {
    constructor() {
        super(...arguments);
        this.name = 'NotFoundError';
        this.code = 404;
    }
}
exports.NotFoundError = NotFoundError;
